ethereum/contracts/ERC20AtomicSwapper.sol

```diff
@@ -0,0 +1,235 @@
+        // Complete the swap.
+        swaps[_secretHashLock].secretKey = _secretKey;
+        swapStates[_secretHashLock] = States.COMPLETED;
+
+        // Pay erc20 token to receiver
+        require(ERC20(ERC20ContractAddr).transfer(swaps[_secretHashLock].receiverAddr, swaps[_secretHashLock].erc20Amount), "Failed to transfer locked asset to receiver address");
+
+        // Emit completion event
+        emit SwapCompletion(msg.sender, swaps[_secretHashLock].receiverAddr, _secretHashLock, _secretKey);
+
+        return true;
+    }
+
+    /// @notice Refunds an atomic swap.
+    ///
+    /// @param _secretHashLock The hash of secretKey and timestamp
+    function refund(bytes32 _secretHashLock) external onlyOpenSwaps(_secretHashLock) onlyAfterExpireHeight(_secretHashLock) returns (bool) {
+        // Expire the swap.
+        swapStates[_secretHashLock] = States.EXPIRED;
+
+        // refund erc20 token to swap creator
+        require(ERC20(ERC20ContractAddr).transfer(swaps[_secretHashLock].sender, swaps[_secretHashLock].erc20Amount), "Failed to transfer locked asset to swap creator");
+
+        // Emit expire event
+        emit SwapExpire(msg.sender, swaps[_secretHashLock].sender, _secretHashLock);
+
+        return true;
+    }
+
+    /// @notice query an atomic swap by secretHashLock
+    ///
+    /// @param _secretHashLock The hash of secretKey and timestamp
+    function querySwapByHashLock(bytes32 _secretHashLock) external view returns(uint64 _timestamp, uint256 _expireHeight, uint256 _erc20Amount, uint256 _bep2Amount, address _sender, address _receiver, bytes20 _BEP2Addr, bytes32 _secretKey, States _status) {
+        Swap memory swap = swaps[_secretHashLock];
+        States status = swapStates[_secretHashLock];
+        return (
+            swap.timestamp,
+            swap.expireHeight,
+            swap.erc20Amount,
+            swap.bep2Amount,
+            swap.sender,
+            swap.receiverAddr,
+            swap.BEP2Addr,
+            swap.secretKey,
+            status
+        );
+    }
+
+    /// @notice query an atomic swap by swap index
+    ///
+    /// @param _index The swap index
+    function querySwapByIndex(uint256 _index) external view returns (bytes32 _secretHashLock, uint64 _timestamp, uint256 _expireHeight, uint256 _erc20Amount, uint256 _bep2Amount, address _sender, address _receiver, bytes20 _BEP2Addr, bytes32 _secretKey, States _status) {
+        bytes32 secretHashLock = indexToSecretHashLock[_index];
+        Swap memory swap = swaps[secretHashLock];
+        States status = swapStates[secretHashLock];
+        return (
+            secretHashLock,
+            swap.timestamp,
+            swap.expireHeight,
+            swap.erc20Amount,
+            swap.bep2Amount,
+            swap.sender,
+            swap.receiverAddr,
+            swap.BEP2Addr,
+            swap.secretKey,
+            status
+        );
+    }
+
```

Comment:
the client may need a way to query by address

Reply:
Client can do this by query SwapInitialization events and add address as filter conditions.